crates/fonticulus/src/fontinfo.rs

```diff
@@ -155,3 +156,135 @@ pub fn get_selection(info: &norad::FontInfo) -> u16 {
     };
     int_list_to_num(&selection) as u16
 }
+
+/// implementation based on ufo2ft:
+/// https://github.com/googlefonts/ufo2ft/blob/main/lib/ufo2ft/util.py#l307
+pub fn calc_code_page_ranges(unicodes: &HashSet<u32>) -> Vec<u8> {
+    let mut code_page_ranges = HashSet::new();
+
+    let ascii_range: HashSet<u32> = (0x20..0x7E).collect();
+    let has_ascii = ascii_range.is_subset(&unicodes);
+    let has_lineart = unicodes.contains(&0x2524); // contains '┤'
+
+    let unicodes_contains = | char | {
+        unicodes.contains(&(char as u32))
+    };
+    if unicodes_contains('Þ') && has_ascii {
+        code_page_ranges.insert(0); // Latin 1
+    }
+    if unicodes_contains('Ľ') && has_ascii {
+        code_page_ranges.insert(1); // Latin 2
+    }
+    if unicodes_contains('Б') {
+        code_page_ranges.insert(2); // Cyrillic
+        if unicodes_contains('Ѕ') && has_lineart {
+            code_page_ranges.insert(57); // IBM Cyrillic
+        }
+        if unicodes_contains('╜') && has_lineart {
+            code_page_ranges.insert(49); // MS-DOS Russian
+        }
+    }
```

Comment:
now I'm just having fun, but I would be tempted to at least handle this simple case of "check char and addd to table" with a macro:

*some time later*:

Okay I'm going to not let myself rabbit-hole on this too hard, but as an illustration: with the following macro, I can replace this section with:

```rust
macro_rules! add_page_if_needed {
    // helper for if there's no condition and no block
    ($collection:ident, $ranges:ident, $chr:expr, $page:expr) => {
        add_page_if_needed!($collection, $chr, $page, true, {});
    };
    // helper for if there's a condition but no block
    ($collection:ident, $ranges:ident, $chr:expr, $page:expr, $constraint:expr) => {
        add_page_if_needed!($collection, $ranges, $chr, $page, $constraint, {});
    };
    // the verbose form
    ($collection:ident, $ranges:ident, $chr:expr, $page:expr, $constraint:expr, $block:block) => {
        if $collection.contains(&($chr as u32)) && $constraint {
            $ranges.insert($page);
            $block;
        }
    };
}
```

```suggestion
    add_page_if_needed!(unicodes, code_page_ranges,'Þ', 0, has_ascii);
    add_page_if_needed!(unicodes, code_page_ranges,'Ľ', 1, has_ascii);
    add_page_if_needed!(unicodes, code_page_ranges, 'Б', 2, true, {
        add_page_if_needed!(unicodes, code_page_ranges,'Ѕ', 57, has_lineart);
        add_page_if_needed!(unicodes, code_page_ranges, '╜', 49, has_lineart);
    });

```

In any case: this is provided more for illustrative purposes, and rust macros can be confusing, but I wanted to point this out as a way I would think about this problem. :)